### PR TITLE
feat: add custom step support

### PR DESCRIPTION
**Pull request by WilliamBergamin**
This PR includes all the changes from #986 aims to add support for custom functions in Bolt Python

### Feedback

I'm looking for feedback on
- Usability, is this what we envisioned the developer experience would be?
- Does this aline with the Bolt-js experience?
- Implementation, does this implementation align with bolt principles
- Tests, is this missing tests that could be valuable for the feature

### Testing

[This gist](https://gist.github.com/WilliamBergamin/e7ccc6ad4383b6883f832cbef2b6215a) was created to show how to test these new feature, use it with the following steps

1. Create a new app
    a. Head to https://api.slack.com/apps/?new_app=1
    b. Click "From an app manifest"
    c. Select YAML and paste the above manifest.yml content
    d. Click create button
2. Install the app into org/workspace
    a. Install the app anyways (when the workspace is in an Org, org-wide installation is required to publish functions)
    b. Grab the xoxb- token (Settings > Install App > Bot User OAuth Token)
    c. `export SLACK_BOT_TOKEN=xoxb-...`
3. Set up Socket Mode
    a. Head to Settings > Basic Information > App-Level Tokens on the https://api.slack.com/apps page
    b. Generate a new token with connections:write scope
    c. `export SLACK_APP_TOKEN=xapp-...`
4. Spin up the app
    a. Add `app.py` and `async_app.py` with the above source code
    b. Pull the latest PR branch of bolt python to your local
    c. Run `scripts/build_pypi_package.sh` this will generate a `.whl` in the `dist/` folder
    d. (Optional) In your project set up a venv with `python -m venv .venv` and `source .venv/bin/activate`
    e. `pip install global/path/to/bolt-python/dist/something.whl`
    f. `pip install aiohttp`
    g `python app.py` or `python async_app.py`
5. Add the custom step to a workflow
    a. Open the workflow builder
    b. Create a new workflow with a link trigger
    d. Seach "Hello" in the Steps view on the right side
    e. Add the step to the workflow (you can set any user ID as the input)
    f. Publish the workflow

### Category

* [x] `slack_bolt.App` and/or its core components
* [x] `slack_bolt.async_app.AsyncApp` and/or its core components
* [ ] Adapters in `slack_bolt.adapter`
* [ ] Document pages under `/docs`
* [ ] Others

## Requirements

Please read the [Contributing guidelines](https://github.com/slackapi/bolt-python/blob/main/.github/contributing.md) and [Code of Conduct](https://slackhq.github.io/code-of-conduct) before creating this issue or pull request. By submitting, you are agreeing to those rules.

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/bolt-python/blob/main/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).
* [x] I've run `./scripts/install_all_and_run_tests.sh` after making the changes.

### CODE REVIEW

File: slack_bolt/app/app.py [hunk 1]
```diff
@@ -828,6 +836,51 @@ def __call__(*args, **kwargs):
...
+        self,
+        callback_id: Union[str, Pattern],
+        matchers: Optional[Sequence[Callable[..., bool]]] = None,
+        middleware: Optional[Sequence[Union[Callable, Middleware]]] = None,
+    ) -> Callable[..., Optional[Callable[..., Optional[BoltResponse]]]]:
+        """Registers a new Function listener.
+        This method can be used as either a decorator or a method.
+            # Use this method as a decorator
+            @app.function("reverse")
+            def reverse_string(event, client: WebClient, context: BoltContext):
```
Comment (seratch) [inline slack_bolt/app/app.py:849]: nit:
```suggestion
            def reverse_string(event: dict, client: WebClient, context: BoltContext):
```

File: slack_bolt/app/app.py [hunk 2]
```diff
@@ -828,6 +836,51 @@ def __call__(*args, **kwargs):
...
+                try:
+                    string_to_reverse = event["inputs"]["stringToReverse"]
+                    client.functions_completeSuccess(
+                        function_execution_id=context.function_execution_id,
+                        outputs={"reverseString": string_to_reverse[::-1]},
+                    )
+                except Exception as e:
+                    client.api_call(
```
Comment (seratch) [inline slack_bolt/app/app.py:857]:
```suggestion
```

File: slack_bolt/app/async_app.py [hunk 3]
```diff
@@ -861,6 +869,52 @@ def __call__(*args, **kwargs):
...
+        self,
+        callback_id: Union[str, Pattern],
+        matchers: Optional[Sequence[Callable[..., Awaitable[bool]]]] = None,
+        middleware: Optional[Sequence[Union[Callable, AsyncMiddleware]]] = None,
+    ) -> Callable[..., Optional[Callable[..., Awaitable[BoltResponse]]]]:
+        """Registers a new Function listener.
+        This method can be used as either a decorator or a method.
+            # Use this method as a decorator
+            @app.function("reverse")
+            async def reverse_string(event, client: AsyncWebClient, complete: AsyncComplete):
```
Comment (seratch) [inline slack_bolt/app/async_app.py:882]:
```suggestion
            async def reverse_string(event: dict, client: AsyncWebClient, complete: AsyncComplete):
```

File: slack_bolt/context/complete/complete.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    def __init__(
+        self,
+        client: WebClient,
+        function_execution_id: Optional[str],
+    ):
+        self.client = client
+        self.function_execution_id = function_execution_id
+
+    def __call__(self, outputs: Dict[str, Any] = {}) -> SlackResponse:
```
Comment (seratch) [inline slack_bolt/context/complete/complete.py:19]: If you need a default value here, you should not use an empty dict value for argument default. See https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python for more details
```suggestion
    def __call__(self, outputs: Optional[Dict[str, Any]] = None) -> SlackResponse:
```
And the following lines of code should be updated accordingly.

If the default value is not necessary, you can simply remove it.

File: slack_bolt/context/fail/async_fail.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+    def __init__(
+        self,
+        client: AsyncWebClient,
+        function_execution_id: Optional[str],
+    ):
+        self.client = client
+        self.function_execution_id = function_execution_id
+
+    async def __call__(self, error: str = "") -> AsyncSlackResponse:
```
Comment (seratch) [inline slack_bolt/context/fail/async_fail.py:19]: Do we want to allow no error here? I believe we don't need to have the default value here.

**Review by seratch — COMMENTED** [review]
Left a few comments

File: slack_bolt/context/complete/complete.py (hunk 4 above)
Comment (WilliamBergamin) [reply slack_bolt/context/complete/complete.py:19]: Wow TIL, did not think the default value would be assigned to the function itself 😮

File: slack_bolt/context/fail/async_fail.py (hunk 5 above)
Comment (WilliamBergamin) [reply slack_bolt/context/fail/async_fail.py:19]: I agree we don't need the default value here 👍

**Comment by sharvesh06** [comment]
@seratch @WilliamBergamin When will this PR be merged?

**Comment by WilliamBergamin** [comment]
@sharvesh06 yess this PR will be merged soon!

File: slack_bolt/app/app.py [hunk 6]
```diff
@@ -178,6 +180,8 @@ def message_hello(message, say):
             url_verification_enabled: False if you would like to disable the built-in middleware (Default: True).
                 `UrlVerification` is a built-in middleware that handles url_verification requests
                 that verify the endpoint for Events API in HTTP Mode requests.
+            attaching_function_token_enabled: False if you would like to disable the built-in middleware (Default: True).
+                `AttachingFunctionToken` is a built-in middleware that handles tokens with function requests from Slack.
```
Comment (filmaj) [inline slack_bolt/app/app.py:184]: I think we should explain what "handling tokens" means in this context. Does the following suggestion make sense?
```suggestion
                `AttachingFunctionToken` is a built-in middleware that uses just-in-time workflow-execution-scoped tokens when your app receives a `function_executed` event.
```

File: slack_bolt/app/app.py [hunk 7]
```diff
@@ -853,6 +861,50 @@ def __call__(*args, **kwargs):
...
+        middleware: Optional[Sequence[Union[Callable, Middleware]]] = None,
+    ) -> Callable[..., Optional[Callable[..., Optional[BoltResponse]]]]:
+        """Registers a new Function listener.
+        This method can be used as either a decorator or a method.
+            # Use this method as a decorator
+            @app.function("reverse")
+            def reverse_string(event: dict, client: WebClient, context: BoltContext):
+                try:
+                    string_to_reverse = event["inputs"]["stringToReverse"]
+                    client.functions_completeSuccess(
```
Comment (filmaj) [inline slack_bolt/app/app.py:877]: Should this example use the client directly, or the `complete` / `fail` shorthand parameter callbacks?

File: slack_bolt/context/base_context.py [hunk 8]
```diff
@@ -103,13 +108,28 @@ def matches(self) -> Optional[Tuple]:
         """Returns all the matched parts in message listener's regexp"""
         return self.get("matches")
 
+    @property
+    def function_execution_id(self) -> Optional[str]:
+        """The `function_execution_id` associated with this request. Only available for function related events"""
```
Comment (filmaj) [inline slack_bolt/context/base_context.py:113]:
```suggestion
        """The bot token resolved for this function request. Only available for `function_executed` and interactivity events scoped to a custom step."""
```

File: slack_bolt/context/base_context.py [hunk 9]
```diff
@@ -103,13 +108,28 @@ def matches(self) -> Optional[Tuple]:
...
+        return self.get("function_execution_id")
+
+    @property
+    def inputs(self) -> Optional[Dict[str, Any]]:
+        """The `inputs` associated with this request. Only available for function related events"""
```
Comment (filmaj) [inline slack_bolt/context/base_context.py:118]:
```suggestion
        """The bot token resolved for this function request. Only available for `function_executed` and interactivity events scoped to a custom step."""
```

File: slack_bolt/context/base_context.py [hunk 10]
```diff
@@ -103,13 +108,28 @@ def matches(self) -> Optional[Tuple]:
...
     # --------------------------------
 
     @property
     def authorize_result(self) -> Optional[AuthorizeResult]:
         """The authorize result resolved for this request."""
         return self.get("authorize_result")
 
+    @property
+    def function_bot_access_token(self) -> Optional[str]:
+        """The bot token resolved for this function request. Only available for function related events"""
```
Comment (filmaj) [inline slack_bolt/context/base_context.py:130]:
```suggestion
        """The bot token resolved for this function request. Only available for `function_executed` and interactivity events scoped to a custom step."""
```

File: slack_bolt/request/async_internals.py [hunk 11]
```diff
@@ -41,6 +44,15 @@ def build_async_context(
     channel_id = extract_channel_id(body)
     if channel_id:
         context["channel_id"] = channel_id
+    function_execution_id = extract_function_execution_id(body)
+    if function_execution_id:
+        context["function_execution_id"] = function_execution_id
+        function_bot_access_token = extract_function_bot_access_token(body)
+        if function_bot_access_token is not None:
```
Comment (filmaj) [inline slack_bolt/request/async_internals.py:51]: Perhaps we should throw an exception if we get here and there is no bot access token?

File: slack_bolt/request/async_internals.py [hunk 12]
```diff
@@ -41,6 +44,15 @@ def build_async_context(
...
+            context["function_bot_access_token"] = function_bot_access_token
+        function_inputs = extract_function_inputs(body)
+        if function_inputs is not None:
```
Comment (filmaj) [inline slack_bolt/request/async_internals.py:54]: I wonder how this code would react to a custom step invocation that accepts zero inputs 🤔

**Review by filmaj — APPROVED** [review]
LGTM! Left some mostly comment-related nits and a few musings about edge cases. But, I ran through the new-app and existing-app tutorials and confirmed everything works as expected 🎉

File: slack_bolt/app/app.py (hunk 7 above)
Comment (WilliamBergamin) [reply slack_bolt/app/app.py:877]: Good call 💯  I think `complete` / `fail` shorthand parameter are better

This example it good to  show how to hand roll completing a function 🤔 not sure where we could include it

Comment (filmaj) [reply slack_bolt/app/app.py:877]: I think just explaining what `complete`/`fail` do, and which underlying HTTP API methods they invoke, is sufficient. Link to the relevant methods and developers will figure the rest out.

File: slack_bolt/context/base_context.py (hunk 9 above)
Comment (WilliamBergamin) [reply slack_bolt/context/base_context.py:118]: I don't think this is the `bot token` I think this is the inputs to the function 🤔

File: slack_bolt/context/base_context.py (hunk 8 above)
Comment (WilliamBergamin) [reply slack_bolt/context/base_context.py:113]: I don't think this is the bot token I think this is the the `function execution id` of the step 🤔

File: slack_bolt/context/base_context.py (hunk 9 above)
Comment (filmaj) [reply slack_bolt/context/base_context.py:118]: o wtf. this might have been a copy-paste error. my mistake!

Comment (filmaj) [reply slack_bolt/context/base_context.py:118]: Edited the suggestion above!

File: slack_bolt/request/async_internals.py (hunk 11 above)
Comment (WilliamBergamin) [reply slack_bolt/request/async_internals.py:51]: We do not need to through an exception here 🤔  this logic extracts the `function_bot_access_token` if it is found in the payload, if it isn't then the payload is not a custom step related requests so nothing happens

File: slack_bolt/request/async_internals.py (hunk 12 above)
Comment (WilliamBergamin) [reply slack_bolt/request/async_internals.py:54]: From my testing a custom step execution payload with no input parameter will have the value `{}` (empty dictionary)
`function_inputs` will never be `None` when handling custom step related events 🟢

**Review by filmaj — APPROVED** [review]

**Pull request merged**